### PR TITLE
Iothub delivery notification

### PR DESCRIPTION
**Pull request by villepalo**
Here is what I was talking about in #391 ('Message Delivered'-notification from IotHub). This feature is mandatory for us, so I just implemented it. If you are interested in having this kind of feature, here it is. If this is not useful to you, feel free to say no.

### CODE REVIEW

File: modules/iothub/src/iothub.c [hunk 1]
```diff
@@ -759,8 +913,40 @@ static void IotHub_Receive(MODULE_HANDLE moduleHandle, MESSAGE_HANDLE messageHan
...
+
+                                    if (mallocAndStrcpy_s(&(userContextCallback->iotHubMessageId), iotHubMessageId) != 0)
+                                    {
+                                        /*Codes_SRS_IOTHUBMODULE_99_008: [ If memory allocation fail when handling "iotHubMessageId" property, `IoTHubClient_SendEventAsync` returns without sending the message ]*/
+                                        LogError("Failed to allocate/copy iotHubMessageId");
+                                        free(userContextCallback);
+                                        IoTHubMessage_Destroy(iotHubMessage);
+                                        ConstMap_Destroy(properties);
+                                        return;
+                                    }
```
Comment (darobs) [inline modules/iothub/src/iothub.c]: Do we want to return here (and in the section above) or make a best effort at sending the message to IoTHub?

File: modules/iothub/src/iothub.c [hunk 2]
```diff
@@ -686,6 +701,145 @@ static IOTHUB_MESSAGE_HANDLE IoTHubMessage_CreateFromGWMessage(MESSAGE_HANDLE me
     return result;
 }
 
+static void SendEventAsync_receiveMessageConfirmation(IOTHUB_CLIENT_CONFIRMATION_RESULT result, void* userContextCallback)
+{
+    if (!userContextCallback)
+    {
+        LogError("Context was null");
+        return;
```
Comment (darobs) [inline modules/iothub/src/iothub.c]: >return [](start = 8, length = 6)

The coding standards for our C code is to have one return call.

File: modules/iothub/src/iothub.c (hunk 1 above)
Comment (villepalo) [reply modules/iothub/src/iothub.c]: Good question. If malloc fails how likely it is that sending message will success? If user needs a notification about delivery, and doesn't get one, wouldn't that be a huge problem too?

File: modules/iothub/src/iothub.c (hunk 2 above)
Comment (villepalo) [reply modules/iothub/src/iothub.c]: fixed

File: modules/iothub/src/iothub.c (hunk 1 above)
Comment (darobs) [reply modules/iothub/src/iothub.c]: The most likely way is if the "iotHubMessageId" string gets trashed in memory and the malloc_and_Strcpy_s fails, but even then it's unlikely that the IoTHubClient_SendEventAsync will succeed as well.  

So it's pretty unlikely it worth trying to send the message. I'm OK with leaving it alone, just more curious if you had a different answer..

**Pull request merged**